File: Data Structures & Algorithms/count-number-of-islands/submission-4.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        
        R = grid.size();
        C = grid[0].size();
        visited = vector<vector<int>>(R, vector<int>(C, 0));

        this->grid = grid;
        this->visited = visited;

        int res = 0;
        for(int r = 0; r < R; r++){
            for(int c = 0; c < C; c++){
                if (grid[r][c] == '1' && visited[r][c] == 0){
                    res++;
                    dfs(r, c);
                }
            }
        }

        return res;
    }
private:
    vector<vector<int>> visited;
    vector<vector<char>> grid;

    int R, C;
    
    bool valid(int r, int c){
        if (r < 0 || r >= R || c < 0 || c >= C || visited[r][c] == 1 || grid[r][c] == '0') return false;
        return true;
    }

    void dfs(int r, int c){
        if (!valid(r, c)) return;
        visited[r][c] = 1;
        dfs(r + 1, c);
        dfs(r - 1, c);
        dfs(r, c + 1);
        dfs(r, c - 1);
    }

};

File: Data Structures & Algorithms/combination-target-sum/submission-4.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = []
        nums.sort()

        def dfs(cur, total, i):
            if total == target:
                res.append(cur.copy())
                return
            if i >= len(nums) or total > target:
                return
            
            cur.append(nums[i])
            dfs(cur, total + nums[i], i)
            cur.pop()
            dfs(cur, total, i + 1)
        dfs([], 0, 0)
        return res

        